# Master's Thesis: Coding for DNA as a Storage

The repository is dedicated to the [master's thesis](https://github.com/feeka/theses/blob/master/Master's%20Thesis%20-%20DNA%20as%20a%20storage%20Error%20Detection.pdf). This version of the repository is not accurate and is a mere draft. Some software scribbles are assembled and simulated.

## Description

The empirical analysis was conducted on a toy-parameter example, where parameters were assumed to be proportional to real-world parameters. A thesis serves as a comprehensive guide for communication specialists or bioinformaticians on the basics of DNA as storage. This repository serves as a scribble of the ideas put together and a simulation based on randomly generated data. Data is generated by the code and fed to the WorkflowHelper simulator.

## Getting Started

### Installing

Download the zip or clone to a local directory and try to run workflow.py or tests.py, depending on the version, using python3. If it asks for dependencies, install them; or, if you do not want to overcrowd your PC with third-party software, use [@venv](https://www.liquidweb.com/kb/how-to-setup-a-python-virtual-environment-on-windows-10/).

## Documentation

During the thesis, an extensive simulation of DNA as a storage system was developed. It is worth mentioning that the author of the thesis developed the codes mentioned below. In chapter four, we briefly introduced the way codes are organized. The appendix covers the functionality of methods and attributes in the listings format and provides documentation of the code attached to the thesis.

> [!NOTE]  
> The thesis was written in 2020, when the technology and the field were still emerging. We created and simulated everything for the toy example. A brand-new repo for a more grounded version of the project I just started (December 2025) is here: [https://github.com/feeka/dna-storage/](https://github.com/feeka/dna-storage/).

### ecc package

There are two classes in the error control codes package: Encoder and Detector (since the Channel involves biological facets, it is located in the helpers package). Each of those classes has corresponding methods that serve specific purposes, such as error control codes. There are more files, but they are scribbles.

### bioinformatics package

There are two classes in the bioinformatics package: Synthesizer and Sequencer. Each of those classes has corresponding methods that serve a specific purpose in bioinformatics. There are more files, but they are scribbles.

### interfaces

Four files are in the interfaces package: Channel, mapper.py, math_helpers.py, and F_Four. Channel and F_Four are classes; math helper.py and mapper are simply a collection of methods. The channel contains methods that simulate errors during sequencing and distort reads correspondingly. F_Four is a class that simulates F4 by overriding standard operators (plus, minus, multiply, divide, and equals). The implementation is straightforward: a dictionary/map with the operation and the resulting answer as a string (for simplicity). The file mapper.py is a Python script that contains two methods to map A, C, T, G < âˆ’ > 0,1,2,3. The file math_helpers contains methods necessary for performing sophisticated mathematical operations.

### workflowmanager.py

WorkflowManager is a Python class that collects the system and executes operations by collecting all the components. The class has eight methods corresponding to DNA steps as a storage schema. 

## Author

M.Sc. Fikrat Talibli

* Email: fikrattalibli@gmail.com
* LinkedIn: [@feeka](https://www.linkedin.com/in/feeka/)

If you have any interest in cleaner code, subsequent versions used in the thesis, or possible collaboration, please do not hesitate to contact me.

## Version History

* 0.1
    * Initial Release
